forem/forem
# Automatically log a creator in after they confirm their email

## What type of PR is this? (check all applicable)


## Description

**A short description of the feature:** 
Right now, when admins click the confirmation link in the email, they land on the login page and have to log in despite having just created their admin account. Instead of forcing the admin to log in here, we should log them in automatically from the moment that they click on the confirmation link in the email.


**How creator signup works:** 
There are two paths for the feature, the one path is when SMTP is enabled and another is when SMTP is not enabled. 

When SMTP is enabled, the workflow is as follows:
1. The creator will signup 
2. They will receive a confirmation email and also be redirected the the /confirm-email path that will say "Great! Now confirm your email address. We've sent an email to abc@gmail.com. Click the button inside to confirm your email." 
3. Once they open up their mail and click on confirm email, they will be automatically signed in.

(3) is the functionality we've implemented in this PR. See gif in QA section. 

When SMTP is not enabled, the workflow is as follows:
1. The creator will signup 
2. The confirmation will be skipped
3. The creator will be automatically logged in. 
Nothing has. changed in this workflow.

## Related Tickets & Documents
Issue: https://github.com/forem/forem/issues/14652
RFC: https://github.com/forem/rfcs/pull/190

## QA Instructions, Screenshots, Recordings

To test both these paths manually do the ffg: 

When SMTP is enabled prior to the creator signing up (this is probably a Forem Cloud use case)
1. Enable SMTP by making sure that there is a value for ENV["SENDGRID_API_KEY"]
2. Restart your server
3. Destroy your database or run ` User.destroy_all`, `Settings::General.waiting_on_first_user = true`, and `Settings::General.mascot_user_id = nil`. 
4. If you'd like to see a preview of the confirmation email in your rails console change [this line](https://github.com/forem/forem/blob/main/app/models/user.rb#L507) to `deliver_now` in the codebase. You can also just find the confirmation token in your rails console as it gets logged and then construct the URL manually `http://localhost:3000/users/confirmation?confirmation_token=<confirmation token>`.

If you go with a preview of the confirmation mail, you should see something like this: 

<img width="1171" alt="Screenshot 2021-09-29 at 18 51 01" src="https://user-images.githubusercontent.com/2786819/135313812-1716810c-dae7-4ee7-9629-095726e2be03.png">

If you're looking for the token, you will see something like this: 

<img width="1639" alt="Screenshot 2021-09-29 at 17 42 13" src="https://user-images.githubusercontent.com/2786819/135313780-000f5e98-8dab-465b-b480-26e30349b2c8.png">

4. Navigate to the confirmation url and the link should automatically log you in. 


![user](https://user-images.githubusercontent.com/2786819/135317081-81855b7e-8d42-4746-b161-84429ee4e148.gif)

Note: In the GIF I just manually override the methods  `smtp_enabled?` to test more easily.

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [x] No, and this is why: I have a test that will partially test this use case, but it still needs some tweaking. We [have logged an issue for it](https://github.com/forem/forem/issues/14655). In addition, in cypress it's difficult to test for conditionals that are dependant on rails environment variables.   

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/o0vwzuFwCGAFO/giphy.gif?cid=ecf05e4756ncfpefyh3kv8fwsfsszdp8wym888xa5mz83eaq&rid=giphy.gif&ct=g)
